### ISSUE TITLE
Gevent compatibility?

### ISSUE THREAD

**Issue by jeffwidman** [confluentinc/confluent-kafka-python#612]
Are there any known limitations with using confluent-kafka-python in a gevent-based application? I searched but couldn't find anything gevent-specific.

In particular, I'm wondering:
* How the callout to `librdkafka`'s pure-C lib interacts with the gevent loop + GIL... will there be weird out of order stuff due to gevent monkey-patching?
* Can the background heartbeat get starved and cause an unnecessary consumer rebalance? Scenario described in https://github.com/dpkp/kafka-python/issues/1515#issuecomment-397147496

related #100, #185

**Comment by lambdaq** [comment]
Looks like no

> https://tech.wayfair.com/2018/07/blocking-io-in-gunicorn-gevent-workers/
> To test this we set up the Flask app in a local container and wrote a harness to send data to the Kafka endpoint concurrently with 100 threads. It turns out that after four concurrent requests (which was also the amount of Gevent workers), the container would stop responding to all requests. The app could only handle four concurrent requests at a time, which was the smoking gun revealing that the Gevent event loop was getting blocked.
> We deployed a fix which changed the code to use `producer.poll(timeout=0)` instead of `producer.flush()`. Response times on the routes that wrote to Kafka lowered from 200ms to 5ns. The Flask servers started behaving normally and were able to handle thousands of requests in parallel.

**Comment by playablebox** [comment]
Any update on this?
The consumer code is also blocking other gevent tasks. 
I tried adding a zero timeout to the consumer as suggested in the Wayfair article `consumer.poll(timeout=0)` but no luck:

   ```
 while True:
        msg = self.consumer.poll(timeout=0)

        if msg is None:
            continue
        if msg.error():
            print("Consumer error: {}".format(msg.error()))
            continue

        print('Received message: {}'.format(msg.value().decode('utf-8')))
        self.consumer.commit()           

    print("closing consumer")
    self.consumer.close()
```

**Comment by mhowlett** [comment]
`commit` is also blocking - you'll need to run it on another thread to not block the event loop. We have a somewhat related blog post coming out soon (producer side) which might help. Companion code: https://github.com/confluentinc/confluent-kafka-python/pull/716

**Comment by mhowlett** [comment]
@jeffwidman - there shouldn't be any issues using with gevent. heartbeats are sent on a background librdkafka thread and aren't affected at all by the GIL. You just need to call poll within the constraints specified by `max.poll.interval.ms`.

**Comment by arjunsingri** [comment]
@mhowlett can you please help me and provide a link to the blog post you mentioned above?

**Comment by yechiel-optimalq** [comment]
@mhowlett - Can you explain what do you mean "there shouldn't be any issues using with gevent. heartbeats are sent on a background librdkafka thread and aren't affected at all by the GIL. " 

If the librdkafka thread is a gevent thread (because of monkey patching) wouldn't it be blocked by other gevent threads?

**Comment by arjunsingri** [comment]
> @mhowlett - Can you explain what do you mean "there shouldn't be any issues using with gevent. heartbeats are sent on a background librdkafka thread and aren't affected at all by the GIL. "
> 
> If the librdkafka thread is a gevent thread (because of monkey patching) wouldn't it be blocked by other gevent threads?

@yechiel-optimalq 

librdkafka is a C library and gevent monkey patching doesn't affect it.

**Comment by yechiel-optimalq** [comment]
Thanks @arjunsingri , We are using confluent-python with gevent and have issues with the heartbeat..  (we work with confluent cloud) it seems that we only heartbeat every half an hour sometimes.

so far we suspected Gevent for that, but if thats not the case we really unsure what can be :\

**Comment by yechiel-optimalq** [comment]
We can report that if we use Gevent.spawn() to spawn a consumer thread , we are experiencing heartbeat issues
Would appreciate any thoughts

**Comment by edenhill** [comment]
Please provide logs of the errors, preferably with `debug=broker` enabled.

**Comment by NadavkOptimalQ** [comment]
[logs.txt](https://github.com/confluentinc/confluent-kafka-python/files/4673057/logs.txt)
Hi @edenhill 
Following your comment to @yechiel-optimalq  I've attached logs we took with debug=all

If this logs is too much and you want only the broker let me know and I will capture again

**Comment by edenhill** [comment]
@NadavkOptimalQ I think the disconnects that you are seeing every 10 minutes is due to how the Azure load balancers close idle connections, see more info here:
https://github.com/edenhill/librdkafka/issues/2845#issuecomment-625433415

As for heartbeat failures, the ones that are seen in your log file is due to the group rebalancing but we can't really see why because the log was captured with debug=broker rather than debug=all

**Comment by NadavkOptimalQ** [comment]
@edenhill 
Thanks for the reply.
I looked at the thread you linked and saw some suggested configuration to handle azure issues but for what I understand they didn't work.

I also saw you talking about version 1.5.0. Is it going to address these issues?

As for the logs I added they were running with debug=all.

Do you have any suggestion for us how to deal with this issue? Do we need to wait for the new version? Or is the new version not going to address this? 

Thank you

**Comment by edenhill** [comment]
> I also saw you talking about version 1.5.0. Is it going to address these issues?

In v1.5.0 we'll try to reuse the least idle connection to allow truly idle connections to time out while keeping at least one known good connection.

> As for the logs I added they were running with debug=all.

`%7|1588062717.229|INIT|kharon-v1-8676b4784f-ctmkv#consumer-2| [thrd:app]: librdkafka v1.3.0 (0x10300ff) kharon-v1-8676b4784f-ctmkv#consumer-2 initialized (builtin.features gzip,snappy,ssl,sasl,regex,lz4,sasl_plain,sasl_scram,plugins,zstd,sasl_oauthbearer, GCC GXX PKGCONFIG INSTALL GNULD LDS LIBDL PLUGINS ZLIB SSL ZSTD HDRHISTOGRAM SNAPPY SOCKEM SASL_SCRAM SASL_OAUTHBEARER CRC32C_HW, debug 0x2)
`

`debug 0x2` = broker

> Do you have any suggestion for us how to deal with this issue? Do we need to wait for the new version? Or is the new version not going to address this?



I'm not sure what the issue actually is, the idle connection closes are not a problem in itself.
Would need the full debug log with `debug=all` (please verify this is set) and a more detailed problem description of what you are seeing, or not seeing.

**Comment by MSeal** [comment]
Closing for now as this is a 5 year old issue without clear next steps. If this is still occurring in a similar context please open a new issue with context.
